### PR TITLE
po/mask manager mode - ensure state stays meaningful

### PR DESCRIPTION
**Pull request by TurboGit**
Multiple fixes for #10770.

- When moving-up or moving-down shapes in the mask manager the new first node must have a clear state and the second one is reset to UNION.

- Disable move up/down depending on the location of the node.

- Disable mode selection for first item in a group.

### CODE REVIEW

File: src/libs/masks.c [hunk 1]
```diff
@@ -737,16 +737,58 @@ static void _tree_moveup(GtkButton *button, dt_lib_module_t *self)
...
     GtkTreePath *item = (GtkTreePath *)items_iter->data;
     GtkTreeIter iter;
+    GtkTreeIter prev_iter;
     if(gtk_tree_model_get_iter(model, &iter, item))
     {
       int grid = -1;
       int id = -1;
       _lib_masks_get_values(model, &iter, NULL, &grid, &id);
 
+      prev_iter = *gtk_tree_iter_copy(&iter);
```
Comment (dterrahe) [inline src/libs/masks.c:754]: Looks like a helper function could avoid some duplication with almost same in `_tree_movedown` below?

**Comment by dterrahe** [comment]
Just a thought from looking at the PR but not any of the code, I wonder if it might not work "better" to for the first mask only _hide_ the state when displaying, set it to UNION by default and ignore it (which already is happening, presumably). That way, you avoid the situation you'd encounter here where if you move the second item to first and then back, its state gets changed to UNION even if it was something else before.

You'd have to deal with legacy data of course where the first mask may still have NONE and moving it down would expose that. And you might have to fix/adjust _many_ other places, so might be less practical as a minor bug fix.

File: src/libs/masks.c (hunk 1 above)
Comment (TurboGit) [reply src/libs/masks.c:754]: Yes, you're certainly right. I should have set this to draft anyway as I also need to free iter/path memory.

**Comment by TurboGit** [comment]
> You'd have to deal with legacy data of course where the first mask may still have NONE and moving it down would expose that. And you might have to fix/adjust _many_ other places, so might be less practical as a minor bug fix.

Probably yes. Not something I'm motivated doing at this point.

**Comment by dterrahe** [comment]
> > You'd have to deal with legacy data of course where the first mask may still have NONE and moving it down would expose that. And you might have to fix/adjust _many_ other places, so might be less practical as a minor bug fix.
> 
> Probably yes. Not something I'm motivated doing at this point.

You would have a much better idea than me, having looked at this just now, but I mention it because there is a _chance_ that none of the extra complexity/loops over all shapes to adjust them are necessary at all and that at just about any point after loading NONE can be changed to UNION by default and in `_lib_masks_list_recurs` for the first item in a group `grpt->state` ignored/overridden with NONE. Once the complexity is there, it is unlikely somebody in future will see through it and recognise it might not be needed...

**Comment by TurboGit** [comment]
@dterrahe : The last version is keeping the operator of the mask and only hide it when it is the first one in the group. I have also done a refactoring. Should be better now.

**Comment by dterrahe** [comment]
Should work fine. Possibly more complicated than needed; `DT_MASKS_STATE_SHOW` is only used during population of the treeview and then it is trivial to set it to false when starting group and true after adding first item in group. So then you don't need to adjust it while manipulating order. But personal preference so fully up to you how you like to approach this of course. And you've probably spent enough time on it already anyway :-)

**Pull request merged**